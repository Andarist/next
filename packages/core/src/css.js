// @flow
import type { CSSCache, Interpolation, ScopedInsertableStyles } from './types'
import { hashString } from 'emotion-utils'
import { handleInterpolation, labelPattern } from './serialize'

export function css(
  strings: Interpolation | string[],
  ...interpolations: Interpolation[]
): ScopedInsertableStyles & { toString: () => string } {
  let stringMode = true
  let styles: string = ''
  let identifierName = ''

  if (strings == null || strings.raw === undefined) {
    stringMode = false
    styles += handleInterpolation.call(this, strings)
  } else {
    styles += strings[0]
  }

  interpolations.forEach(function(interpolation, i) {
    styles += handleInterpolation.call(this, interpolation)
    if (stringMode === true && strings[i + 1] !== undefined) {
      styles += strings[i + 1]
    }
  }, this)
  styles = styles.replace(labelPattern, (match, p1: string) => {
    identifierName += `-${p1}`
    return ''
  })
  let name

  const ret: Object = {
    styles,
    type: 1,
    get name() {
      delete this.name
      this.name = hashString(styles) + identifierName
      return this.name
    },
    toString() {
      return `css-${this.name}`
    }
  }
  return ret
}
