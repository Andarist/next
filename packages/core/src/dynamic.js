// @flow
import * as React from 'react'
import { hydration, isBrowser, insertStyles } from './utils'
import { serializeStyles } from './serialize'
import { CSSContext } from './context'
import { DynamicStyleSheet } from './sheet'

type Props = {
  css: Object,
  render: string => React.Node
}

const StyleSheetOptions = { key: 'dynamic' }

export class Dynamic extends React.Component<Props> {
  shouldHydrate: boolean
  serialized: string
  shouldHydrate = hydration.shouldHydrate
  sheet: DynamicStyleSheet
  componentDidMount() {
    hydration.shouldHydrate = false
  }
  componentWillUnmount() {
    if (this.sheet !== undefined) {
      this.sheet.flush()
    }
  }
  render() {
    return (
      <CSSContext.Consumer>
        {context => {
          const { css } = this.props
          const serialized = serializeStyles([css])
          const rules = context.stylis(
            `.css-${serialized.name}`,
            serialized.styles
          )
          if (this.sheet === undefined && isBrowser) {
            this.sheet = new DynamicStyleSheet(StyleSheetOptions)
            this.sheet.inject()
          }
          if (isBrowser) {
            this.sheet.insertRules(rules)
          }

          if (
            this.serialized === undefined &&
            (this.shouldHydrate || !isBrowser)
          ) {
            this.serialized = rules.join('')
          }
          const child = this.props.render(serialized.cls)
          if (this.shouldHydrate || !isBrowser) {
            return (
              <React.Fragment>
                <style
                  data-more=""
                  dangerouslySetInnerHTML={{ __html: this.serialized }}
                />
                {child}
              </React.Fragment>
            )
          }
          return child
        }}
      </CSSContext.Consumer>
    )
  }
}
