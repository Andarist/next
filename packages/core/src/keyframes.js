// @flow
import * as React from 'react'
import { CSSContext } from './context'
import type { CSSContextType, InsertableStyles } from './types'
import { hydration, isBrowser, insertStyles } from './utils'
import { serializeStyles } from './serialize'

export const keyframes = (arg: { styles: string, name: string }) => {
  const name = `animation-${arg.name}`
  const ret = {
    styles: `@keyframes ${name}{${arg.styles}}`,
    name,
    scope: '',
    toString: () => `animation-${arg.name}`
  }
  return ret
}
