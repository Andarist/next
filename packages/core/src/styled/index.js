// @flow
import * as React from 'react'
import { STYLES_KEY } from 'emotion-utils'
import type { ElementType } from 'react'
import {
  testOmitPropsOnComponent,
  testAlwaysTrue,
  testOmitPropsOnStringTag,
  omitAssign,
  tags,
  type Interpolations,
  type StyledOptions,
  type CreateStyled
} from './utils'
import {
  getRegisteredStyles,
  insertStyles,
  hydration,
  isBrowser
} from '../utils'
import { serializeStyles } from '../serialize'
import { CSSContext } from '../context'

let createStyled: CreateStyled = (tag: any, options?: StyledOptions) => {
  if (process.env.NODE_ENV !== 'production') {
    if (tag === undefined) {
      throw new Error(
        'You are trying to create a styled element with an undefined component.\nYou may have forgotten to import it.'
      )
    }
  }
  let identifierName
  if (options !== undefined) {
    identifierName = options.label
  }
  const isReal = tag.__emotion_real === tag
  const baseTag = (isReal && tag.__emotion_base) || tag

  const omitFn =
    typeof baseTag === 'string' &&
    baseTag.charAt(0) === baseTag.charAt(0).toLowerCase()
      ? testOmitPropsOnStringTag
      : testOmitPropsOnComponent

  return function() {
    let args = arguments
    let styles =
      isReal && tag[STYLES_KEY] !== undefined ? tag[STYLES_KEY].slice(0) : []
    if (identifierName !== undefined) {
      styles.push(`label:${identifierName};`)
    }
    if (args[0] == null || args[0].raw === undefined) {
      styles.push.apply(styles, args)
    } else {
      styles.push(args[0][0])
      let len = args.length
      let i = 1
      for (; i < len; i++) {
        styles.push(args[i], args[0][i])
      }
    }

    class Styled extends React.Component<*> {
      mergedProps: Object
      shouldHydrate: boolean
      serialized: string
      static toString: () => string
      static __emotion_real: any
      static __emotion_styles: Interpolations
      static __emotion_base: Styled
      static withComponent: (ElementType, options?: StyledOptions) => any
      constructor(props) {
        super(props)
        this.shouldHydrate = hydration.shouldHydrate
      }
      componentWillUnmount() {
        hydration.shouldHydrate = false
      }
      render() {
        return (
          <CSSContext.Consumer>
            {context => {
              let className = ''
              let classInterpolations = []
              this.mergedProps = omitAssign(testAlwaysTrue, {}, this.props, {
                theme: context.theme || this.props.theme || {}
              })
              if (
                this.props.className &&
                typeof this.props.className === 'string'
              ) {
                className += getRegisteredStyles(
                  context.registered,
                  classInterpolations,
                  this.props.className
                )
              }
              const serialized = serializeStyles.call(
                this,
                context.registered,
                styles.concat(classInterpolations)
              )
              const rules = insertStyles(context, serialized)
              className += serialized.toString()

              if (
                this.serialized === undefined &&
                (this.shouldHydrate || !isBrowser)
              ) {
                this.serialized = rules
              }

              const ele = React.createElement(
                baseTag,
                omitAssign(omitFn, {}, this.props, {
                  className,
                  ref: this.props.innerRef
                })
              )
              if (this.shouldHydrate || !isBrowser) {
                return (
                  <React.Fragment>
                    <style
                      data-more={serialized.name}
                      dangerouslySetInnerHTML={{ __html: this.serialized }}
                    />
                    {ele}
                  </React.Fragment>
                )
              }
              return ele
            }}
          </CSSContext.Consumer>
        )
      }
    }
    Styled.displayName =
      identifierName !== undefined
        ? identifierName
        : `Styled(${
            typeof baseTag === 'string'
              ? baseTag
              : baseTag.displayName || baseTag.name || 'Component'
          })`

    Styled[STYLES_KEY] = styles
    Styled.__emotion_base = baseTag
    Styled.__emotion_real = Styled

    Styled.withComponent = (
      nextTag: ElementType,
      nextOptions?: StyledOptions
    ) => {
      return createStyled(
        nextTag,
        nextOptions !== undefined
          ? // $FlowFixMe
            omitAssign(testAlwaysTrue, {}, options, nextOptions)
          : options
      )(...args)
    }

    return Styled
  }
}

tags.forEach(tagName => {
  createStyled[tagName] = createStyled(tagName)
})

export default createStyled
