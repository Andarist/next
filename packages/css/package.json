{
  "name": "@emotion/css",
  "version": "0.2.3",
  "description": "a function to serialize css and object styless",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "repository": "https://github.com/emotion-js/next/tree/master/packages/css",
  "dependencies": {
    "@emotion/serialize": "^0.2.1",
    "@emotion/types": "^0.2.3",
    "emotion-utils": "^9.0.1"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "src",
    "dist"
  ]
}
