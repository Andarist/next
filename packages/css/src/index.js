// @flow
import type { Interpolation, ScopedInsertableStyles } from '@emotion/types'
import hashString from '@emotion/hash'
import { handleInterpolation, labelPattern } from '@emotion/serialize'

function css(
  strings: Interpolation | string[],
  ...interpolations: Interpolation[]
): ScopedInsertableStyles {
  let stringMode = true
  let styles: string = ''
  let identifierName = ''

  if (strings == null || strings.raw === undefined) {
    stringMode = false
    styles += handleInterpolation.call(this, strings)
  } else {
    styles += strings[0]
  }

  interpolations.forEach(function(interpolation, i) {
    styles += handleInterpolation.call(this, interpolation)
    if (stringMode === true && strings[i + 1] !== undefined) {
      styles += strings[i + 1]
    }
  }, this)
  styles = styles.replace(labelPattern, (match, p1: string) => {
    identifierName += `-${p1}`
    return ''
  })
  let name = hashString(styles) + identifierName

  return {
    type: 1,
    name,
    styles
  }
}

export default css
