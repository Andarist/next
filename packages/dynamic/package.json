{
  "name": "@emotion/dynamic",
  "version": "0.0.5",
  "description": "React component to render highly dynamic styles for emotion",
  "main": "dist/index.cjs.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "repository":
    "https://github.com/emotion-js/next/tree/master/packages/dynamic",
  "dependencies": {
    "@emotion/sheet": "^0.0.5",
    "@emotion/utils": "^0.0.5",
    "@emotion/serialize": "^0.0.5",
    "@emotion/types": "^0.0.5"
  },
  "peerDependencies": {
    "@emotion/core": "^0.0.4",
    "react": ">=16.3.0 || 16.3.0-alpha.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": ["src", "dist"]
}
