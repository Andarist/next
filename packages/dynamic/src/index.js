// @flow
import * as React from 'react'
import { consumer } from '@emotion/core'
import { isBrowser, shouldSerializeToReactTree } from '@emotion/utils'
import { serializeStyles } from '@emotion/serialize'
import { DynamicStyleSheet } from '@emotion/sheet'
import type { CSSContextType } from '@emotion/types'

type Props = {
  css: Object,
  render: string => React.Node
}

const StyleSheetOptions = { key: 'dynamic' }

class Dynamic extends React.Component<Props> {
  sheet: DynamicStyleSheet
  static __emotion_component = true
  renderChild = (context: CSSContextType) => {
    const { css } = this.props
    const serialized = serializeStyles([css])
    const rules = context.stylis(`.css-${serialized.name}`, serialized.styles)
    if (this.sheet === undefined && isBrowser) {
      this.sheet = new DynamicStyleSheet(StyleSheetOptions)
      this.sheet.inject()
    }
    if (isBrowser) {
      this.sheet.insertRules(rules)
    }

    const child = this.props.render(serialized.cls)
    if (shouldSerializeToReactTree && rules.length !== 0) {
      return (
        <React.Fragment>
          <style
            data-emotion-ssr={serialized.name}
            dangerouslySetInnerHTML={{ __html: rules.join('') }}
          />
          {child}
        </React.Fragment>
      )
    }
    return child
  }
  componentWillUnmount() {
    if (this.sheet !== undefined) {
      this.sheet.flush()
    }
  }
  render() {
    return consumer(this, this.renderChild)
  }
}

export default Dynamic
