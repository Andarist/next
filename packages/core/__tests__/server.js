/** @jsx jsx
 * @jest-environment node
 */
// @flow
import 'test-utils/no-test-mode'
import * as React from 'react'
import cases from 'jest-in-case'
import jsx from '@emotion/jsx'
import Style from '@emotion/style'
import styled from '@emotion/styled'
import Global from '@emotion/global'
import dynamic from '@emotion/dynamic'
import css from '@emotion/css'
import keyframes from '@emotion/keyframes'
import { renderToString } from 'react-dom/server'
import HTMLSerializer from 'jest-serializer-html'

expect.addSnapshotSerializer(HTMLSerializer)

cases(
  'ssr',
  opts => {
    expect(renderToString(opts.render())).toMatchSnapshot()
  },
  {
    basic: {
      render: () => <div css={{ color: 'hotpink' }}>some hotpink text</div>
    },
    global: {
      render: () => (
        <Global
          css={{
            html: {
              backgroundColor: 'hotpink'
            }
          }}
        />
      )
    },
    keyframes: {
      render: () => {
        const animation = keyframes(
          css`
            from {
              color: green;
            }
            to {
              color: blue;
            }
          `
        )

        return (
          <React.Fragment>
            <Style styles={animation} />
            <div css={{ animation: `1s ${animation.toString()}` }} />
          </React.Fragment>
        )
      }
    },
    dynamic: {
      render: () => {
        const Dynamic = dynamic`
          color: hotpink;
        `
        return (
          <Dynamic
            children={className => {
              return <div className={className} />
            }}
          />
        )
      }
    },
    className: {
      render: () => {
        const className = css`
          color: hotpink;
        `

        return (
          <div>
            <Style styles={className} />
            <div className={className} />
          </div>
        )
      }
    },
    'only render a style once with the css prop': {
      render: () => {
        return (
          <div css={{ color: 'green' }}>
            <div css={{ color: 'hotpink' }} />
            <div css={{ color: 'hotpink' }} />
          </div>
        )
      }
    },
    'only render a style once with Style': {
      render: () => {
        const className = css`
          color: hotpink;
        `
        return (
          // note that there should be a container element with a style/Provider
          // otherwise the style will render twice
          <div css={{ color: 'green' }}>
            <Style styles={className} />
            <Style styles={className} />
            <div className={className} />
          </div>
        )
      }
    },

    'only render a style once with styled': {
      render: () => {
        const SomeComponent = styled.div`
          color: hotpink;
        `
        return (
          <div css={{ color: 'green' }}>
            <SomeComponent />
            <SomeComponent />
          </div>
        )
      }
    }
  }
)
