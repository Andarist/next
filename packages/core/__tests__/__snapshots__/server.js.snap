// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr basic 1`] = `

<style data-emotion-ssr="1lrxbo5">
  .css-1lrxbo5{color:hotpink;}
</style>
<div class="css-1lrxbo5">
  some hotpink text
</div>

`;

exports[`ssr className 1`] = `

<div data-reactroot>
  <style data-emotion-ssr="181qz4b">
    .css-181qz4b{color:hotpink;}
  </style>
  <div class="css-181qz4b">
  </div>
</div>

`;

exports[`ssr dynamic 1`] = `

<style data-emotion-ssr="1lrxbo5">
  .css-1lrxbo5{color:hotpink;}
</style>
<div class="css-1lrxbo5">
</div>

`;

exports[`ssr global 1`] = `

<style data-emotion-ssr="6l7sk3">
  html{background-color:hotpink;}
</style>

`;

exports[`ssr keyframes 1`] = `

<style data-emotion-ssr="animation-1kyaq4y">
  @-webkit-keyframes animation-1kyaq4y{from{color:green;}to{color:blue;}}@keyframes animation-1kyaq4y{from{color:green;}to{color:blue;}}
</style>
<style data-emotion-ssr="q9uc0r">
  .css-q9uc0r{-webkit-animation:1s animation-1kyaq4y;animation:1s animation-1kyaq4y;}
</style>
<div class="css-q9uc0r">
</div>

`;

exports[`ssr only render a style once with Style 1`] = `

<style data-emotion-ssr="bjcoli">
  .css-bjcoli{color:green;}
</style>
<div class="css-bjcoli">
  <style data-emotion-ssr="181qz4b">
    .css-181qz4b{color:hotpink;}
  </style>
  <div class="css-181qz4b">
  </div>
</div>

`;

exports[`ssr only render a style once with styled 1`] = `

<style data-emotion-ssr="bjcoli">
  .css-bjcoli{color:green;}
</style>
<div class="css-bjcoli">
  <style data-emotion-ssr="181qz4b">
    .css-181qz4b{color:hotpink;}
  </style>
  <div class="css-181qz4b">
  </div>
  <div class="css-181qz4b">
  </div>
</div>

`;

exports[`ssr only render a style once with the css prop 1`] = `

<style data-emotion-ssr="bjcoli">
  .css-bjcoli{color:green;}
</style>
<div class="css-bjcoli">
  <style data-emotion-ssr="1lrxbo5">
    .css-1lrxbo5{color:hotpink;}
  </style>
  <div class="css-1lrxbo5">
  </div>
  <div class="css-1lrxbo5">
  </div>
</div>

`;
